py_ase: new versions and maintainer

alikhamze opened the pull request:
Updates:
- Adds versions 3.24.0 and 3.25.0
- Adds myself as a maintainer.
- Use the `when` context manager for dependencies for improved maintainability and readability.
- Raises minimum python version to 3.4 since python 2 is no longer supported (only relevant for the oldest versions)

alikhamze committed:
Add versions 3.24.0 and 3.25.0

alikhamze committed:
Add myself as a maintainer.

alikhamze commented:
This will fix the failing CI in #361 after the latest commits to add GPAW@25.7.0.

tldahlgren commented on repos/spack_repo/builtin/packages/py_ase/package.py:
```diff
@@ -34,7 +38,8 @@ class PyAse(PythonPackage):
     depends_on("python@3.6:", type=("build", "run"), when="@3.20.0:")
-    depends_on("python@3.8:", type=("build", "run"), when="@3.23.0:")
+    depends_on("python@3.8:", type=("build", "run"), when="@3.23.0")
+    depends_on("python@3.9:", type=("build", "run"), when="@3.24.0:")
```
A couple of the dependencies either need lower bounds or updated lower bounds (https://spack.readthedocs.io/en/latest/packaging_guide_creation.html#conditional-dependencies) for the newer version(s) per `pyproject.toml`:

For`3.24.0` and `3.25.0`:

```
[build-system]
requires = ['setuptools >= 61.0.0']
...
dependencies = [
    'numpy>=1.19.5',  # January 2021
...
```

alikhamze replied:
Ah, I was just looking at the release notes and missed those, thanks for catching them! Pushed a fix and switched to using the when context manager for dependencies.

tldahlgren reviewed (changes requested):
Additional dependencies, conditioned on the version, are needed starting with `3.24.0`. See comment(s).

tldahlgren commented:
Very odd that `pypi.org`'s search doesn't appear to yield `ase` though they are hosting the package's files.

alikhamze committed:
Switch to using the `when` context manager for dependencies and update dependecy versions..

Where the pyproject.toml or setup.py did not provide version constraints, I kept the ones from the older spack package.

alikhamze commented:
I'm also not able to find it with pypi's search, that is very odd.

tldahlgren commented:
> Very odd that `pypi.org`'s search doesn't appear to yield `ase` though they are hosting the package's files.

Turns out there's a reindex going on:  https://status.python.org.

tldahlgren commented on repos/spack_repo/builtin/packages/py_ase/package.py:
```diff
@@ -31,15 +35,53 @@ class PyAse(PythonPackage):
+        depends_on("py-matplotlib@3.3.4:", type=("build", "run"))
+        depends_on("py-setuptools@61.0.0:", type="build")
+
+    with when("@3.24.0"):
+        depends_on("python@3.9:", type=("build", "run"))
+        depends_on("py-numpy@1.19.5:", type=("build", "run"))
+        depends_on("py-scipy@1.6.0:", type=("build", "run"))
+        depends_on("py-matplotlib@3.3.4:", type=("build", "run"))
+        depends_on("py-setuptools@61.0.0:", type="build")
+
```
There is no reason to have this when the dependencies in the context above (lines 38-43) cover both new versions that have the same dependencies thanks to the use of the version range specified at line 38.

alikhamze replied:
Ack, my mistake!

tldahlgren commented on repos/spack_repo/builtin/packages/py_ase/package.py:
```diff
@@ -31,15 +35,53 @@ class PyAse(PythonPackage):
+    with when("@3.24.0:"):
+        depends_on("python@3.9:", type=("build", "run"))
+        depends_on("py-numpy@1.19.5:", type=("build", "run"))
+        depends_on("py-scipy@1.6.0:", type=("build", "run"))
```
FWIW. You could drop the `.0` patch release and have the same results (https://spack.readthedocs.io/en/latest/packaging_guide_creation.html#specifying-versions-in-other-directives):

```suggestion
        depends_on("py-scipy@1.6:", type=("build", "run"))
```

alikhamze replied:
Done, will keep this in mind for the future!

tldahlgren commented on repos/spack_repo/builtin/packages/py_ase/package.py:
```diff
@@ -31,15 +35,53 @@ class PyAse(PythonPackage):
+        depends_on("py-numpy@1.19.5:", type=("build", "run"))
+        depends_on("py-scipy@1.6.0:", type=("build", "run"))
+        depends_on("py-matplotlib@3.3.4:", type=("build", "run"))
+        depends_on("py-setuptools@61.0.0:", type="build")
```
Same here for minor and patch versions:

```suggestion
        depends_on("py-setuptools@61:", type="build")
```

tldahlgren commented on repos/spack_repo/builtin/packages/py_ase/package.py:
```diff
@@ -31,15 +35,53 @@ class PyAse(PythonPackage):
+        depends_on("py-matplotlib@3.3.4:", type=("build", "run"))
+        depends_on("py-setuptools@61.0.0:", type="build")
+
+    with when("@3.21.0:3.21.1"):
```
FWIW. *Probably* safe to use the simpler range:

```suggestion
    with when("@3.21"):
```

tldahlgren reviewed (changes requested):
There some duplication and optional simplification. See comments for details.

alikhamze committed:
Remove accidentally duplicated dependencies.

alikhamze committed:
Update homepage and pypi version.

alikhamze committed:
Simplify version constraints.

tldahlgren reviewed (approved):
Thanks.

alikhamze commented:
Thank you for the feedback!
